File: kirun-js/src/engine/function/system/date/LastFirstOf.ts
import { DateTime } from 'luxon';
import { EventResult } from '../../../model/EventResult';
import { FunctionOutput } from '../../../model/FunctionOutput';
import { FunctionSignature } from '../../../model/FunctionSignature';
import { Namespaces } from '../../../namespaces/Namespaces';
import { FunctionExecutionParameters } from '../../../runtime/FunctionExecutionParameters';
import { MapUtil } from '../../../util/MapUtil';
import { AbstractFunction } from '../../AbstractFunction';
import { AbstractDateFunction } from './AbstractDateFunction';
import { Parameter } from '../../../model/Parameter';
import { Schema } from '../../../json/schema/Schema';
import { getDateTime } from './common';

export class LastFirstOf extends AbstractFunction {
    private readonly signature: FunctionSignature;
    private readonly isLast: boolean;

    constructor(isLast: boolean) {
        super();

        this.isLast = isLast;
        this.signature = new FunctionSignature(isLast ? 'LastOf' : 'FirstOf')
            .setNamespace(Namespaces.DATE)
            .setParameters(
                new Map([
                    [
                        AbstractDateFunction.PARAMETER_TIMESTAMP_NAME,
                        new Parameter(
                            AbstractDateFunction.PARAMETER_TIMESTAMP_NAME,
                            Schema.ofRef(Namespaces.DATE + '.Timestamp'),
                        ).setVariableArgument(true),
                    ],
                ]),
            )
            .setEvents(
                new Map([
                    [
                        AbstractDateFunction.EVENT_TIMESTAMP_NAME,
                        AbstractDateFunction.EVENT_TIMESTAMP,
                    ],
                ]),
            );
    }

    public getSignature(): FunctionSignature {
        return this.signature;
    }

    protected internalExecute(context: FunctionExecutionParameters): Promise<FunctionOutput> {
        const timestamps = context
            .getArguments()
            ?.get(AbstractDateFunction.PARAMETER_TIMESTAMP_NAME);

        if (!timestamps?.length) {
            throw new Error('No timestamps provided');
        }

        const dateTimes: DateTime[] = timestamps.map((ts: string) => getDateTime(ts));
        dateTimes.sort((a, b) => a.toMillis() - b.toMillis());

        return Promise.resolve(
            new FunctionOutput([
                EventResult.outputOf(
                    MapUtil.of(
                        AbstractDateFunction.EVENT_TIMESTAMP_NAME,
                        dateTimes[this.isLast ? dateTimes.length - 1 : 0].toISO(),
                    ),
                ),
            ]),
        );
    }
}

File: kirun-py/src/kirun_py/function/system/date/last_first_of.py
from __future__ import annotations

from typing import TYPE_CHECKING

from kirun_py.function.abstract_function import AbstractFunction
from kirun_py.json.schema.schema import Schema
from kirun_py.model.event_result import EventResult
from kirun_py.model.function_output import FunctionOutput
from kirun_py.model.function_signature import FunctionSignature
from kirun_py.model.parameter import Parameter
from kirun_py.namespaces.namespaces import Namespaces
from kirun_py.function.system.date.abstract_date_function import AbstractDateFunction
from kirun_py.function.system.date.common import get_datetime, datetime_to_iso

if TYPE_CHECKING:
    from kirun_py.runtime.function_execution_parameters import FunctionExecutionParameters


class LastFirstOf(AbstractFunction):

    def __init__(self, is_last: bool) -> None:
        super().__init__()
        self._is_last = is_last

        self._signature = (
            FunctionSignature('LastOf' if is_last else 'FirstOf')
            .set_namespace(Namespaces.DATE)
            .set_parameters({
                AbstractDateFunction.PARAMETER_TIMESTAMP_NAME: Parameter(
                    AbstractDateFunction.PARAMETER_TIMESTAMP_NAME,
                    Schema.of_ref(Namespaces.DATE + '.Timestamp'),
                ).set_variable_argument(True),
            })
            .set_events({
                AbstractDateFunction.EVENT_TIMESTAMP.get_name(): AbstractDateFunction.EVENT_TIMESTAMP,
            })
        )

    def get_signature(self) -> FunctionSignature:
        return self._signature

    async def internal_execute(self, context: FunctionExecutionParameters) -> FunctionOutput:
        args = context.get_arguments() or {}
        timestamps = args.get(AbstractDateFunction.PARAMETER_TIMESTAMP_NAME)

        if not timestamps:
            raise ValueError('No timestamps provided')

        datetimes = [get_datetime(ts) for ts in timestamps]
        datetimes.sort()

        chosen = datetimes[-1] if self._is_last else datetimes[0]

        return FunctionOutput([
            EventResult.output_of(
                {AbstractDateFunction.EVENT_TIMESTAMP_NAME: datetime_to_iso(chosen)}
            ),
        ])
